src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-                std::get<1>(_file->meshes.back()) = end;
-
-                /* Save name and offset of the new one. The end offset will be
-                   updated later. */
-                if(!name.empty())
-                    _file->meshesForName.emplace(name, _file->meshes.size());
-                _file->meshNames.emplace_back(std::move(name));
-                _file->meshes.emplace_back(_file->in->tellg(), 0, positionIndexOffset, textureCoordinateIndexOffset, normalIndexOffset);
+            if(sectionPrimitive != keyword[0]) {
+                /* Create new section, mixed primitives! */
+                finishSection();
             }
 
-            continue;
+            if(section.data() == nullptr) {
+                /* Create new section */
+                section = line;
+                sectionPrimitive = keyword[0];
+            }
 
-        /* If there are any data/indices before the first name, it means that
-           the first object is unnamed. We need to check for them. */
+            ++minSectionPrimitives;
 
-        /* Vertex data, update index offset for the following meshes */
-        } else if(keyword == "v") {
-            ++positionIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-        } else if(keyword == "vt") {
-            ++textureCoordinateIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-        } else if(keyword == "vn") {
-            ++normalIndexOffset;
-            thisIsFirstMeshAndItHasNoData = false;
-
-        /* Index data, just mark that we found something for first unnamed
-           object */
-        } else if(thisIsFirstMeshAndItHasNoData) for(const std::string& data: {"p", "l", "f"}) {
-            if(keyword == data) {
-                thisIsFirstMeshAndItHasNoData = false;
-                break;
-            }
+        /* Object name */
+        } else if(keyword == "o") {
+            finishSection();
+            finishObject();
+
+            ArrayView<char> name;
+            std::tie(name, pos) = nextWord(pos);
+
+            _state->objects.emplace_back(new ObjObject{name});
+            object = _state->objects.back().get();
+            group = nullptr;
+            meshData = nullptr;
+
+        /* Object group */
+        } else if(keyword == "g") {
+            //TODO: Handle geometry shared by multiple groups, e.g. g group1 group2
+            ensureObject();
+            object->groups.emplace_back(new ObjGroup{*object});
+            group = object->groups.back().get();
+
+            meshData = nullptr;
+
+            ArrayView<char> name;
+            std::tie(name, pos) = nextWord(pos);
+            group->name = std::string{name.data(), name.size()};
+
+        /* Load a material library */
+        } else if(keyword == "mtllib") {
+            ArrayView<char> word;
+            pos = skipWhitespaces(pos);
+            std::tie(word, pos) = nextWord(pos);
+
+            parseMaterialLibrary(word);
```

Comment:
From what I understand, just opening the file basically parses all of it and bails out if there is an error in any part of the file or if an external `*.mtl` file can't be found, even though I might not ever need that.

Isn't there a way to first just do a quick look over the file, remember mapping between names and IDs and do the whole numeric data parsing only on-demand when a particular mesh/object/material is requested?